### DISCUSSION TITLE
How do I receive the response data in the listening event?

### DISCUSSION THREAD

**Discussion by HelplessMan** [telegram-bot-sdk/telegram-bot-sdk#80] [Q&A] [answered]
>telegram.php

```
'bots' => [
        'default' => [
            'token' => env('TELEGRAM_BOT_TOKEN', 'YOUR-BOT-TOKEN'),

            'commands' => [
                 'start' => App\Telegram\Commands\Start::class,
                 'info' => App\Telegram\Commands\Info::class,
                 'test' => App\Telegram\Commands\TestCommand::class,
            ],

            'listen' => [
                'update' => [],
                'webhook.failed' => [],

                // Example of various events fired.
                'message' => [
                    App\Telegram\Commands\Message::class
                ],
                'poll' => [

                ],
                'message.left_chat_member' => [
                    App\Telegram\Commands\LeftChatMember::class,
                ],
                'message.new_chat_members' => [
                    App\Telegram\Commands\NewChatMember::class,
                ],
                'inline_query.location' => [],
                'callback_query' => [
                    App\Telegram\Commands\CallbackQuery::class,
                ]
            ],
        ]
]
```
>  NewChatMember.php
```
 public function handle(UpdateEvent $updateEvent)
    {

        logger('new chat member event :' . json_encode($updateEvent));
        $this->chatId   = $updateEvent->update->message->chat->id;
        $this->memberId = $updateEvent->update->message->new_chat_member->id ?? '';
        $captchaContent = $this->getCaptchaContent();

        $key  = 'captcha_' . md5($captchaContent['key']);
        $text = implode('', Cache::get($key));

        $captchaUrl = $this->getCaptchaUrl($captchaContent);

        logger('message:' . json_encode($updateEvent, JSON_THROW_ON_ERROR) . ' text:' . $text . ' newMemberId:' . $this->memberId);

        logger('captchaUrl:' . $captchaUrl);
        //todo 
        $caption = 'caption';
        
        $response = Telegram::sendPhoto([
            'chat_id' => $this->chatId,
            'photo' => $captchaUrl,
            'caption' => $caption,
            'reply_markup' => json_encode([
                'inline_keyboard' => Utils::generateCaptchaButton($text)
            ])
        ]);

        logger('newChatMember:' . $response);
    }

```

>Q1:  In the event, I use UpdateEvent to accept the event object. Is it correct to use it like this, but at present, he can get the post parameter of the webhook

>Q2: In the event class, how do I send information and pictures? When I use $this->bot, it prompts me that I cannot find the sendPhoto method. Telegram::sendPhoto returns a null value, and here I need to know his message_id

>Q3: 4. When can the documentation of X be updated...Using it without documentation will greatly affect development efficiency...

**Comment by HelplessMan** [comment]
@irazasyed 


please help me bro... i posted my problem in discussions, no one help me...

**Reply by irazasyed** [reply]
> Q1: In the event, I use UpdateEvent to accept the event object. Is it correct to use it like this, but at present, he can get the post parameter of the webhook

Yes. That's correct. 

Auto: `php artisan telegram:install` this will publish config and scaffolding for you to get started (config, command, listener, env var).

Manual: Just make sure to add `Telegram::listen(true);` in your webhook route. `true` in this case means to process inbound updates via webhook.

> Q2: In the event class, how do I send information and pictures? When I use $this->bot, it prompts me that I cannot find the sendPhoto method. Telegram::sendPhoto returns a null value, and here I need to know his message_id

If you're not using multiple bots, then you can simply do `Telegram::sendPhoto($params)`, just make sure to import the Telegram facade. If you're using multiple bots, then a better way to do it is using the Bot instance in the event. 

Ex: 

```php
$bot = $event->bot;
$bot->sendPhoto($params);
```

Don't worry about not found IDE warnings as these methods are being called through the magic methods dynamically linked from the Api class.

> Q3: 4. When can the documentation of X be updated...Using it without documentation will greatly affect development efficiency...

4.x docs are being worked on, there are no ETAs as we've been refactoring a lot of things internally like the new `command()` methods and the new response objects and so on.

**Reply by HelplessMan** [reply]
> > Q1: In the event, I use UpdateEvent to accept the event object. Is it correct to use it like this, but at present, he can get the post parameter of the webhook
> 
> Yes. That's correct.
> 
> Auto: `php artisan telegram:install` this will publish config and scaffolding for you to get started (config, command, listener, env var).
> 
> Manual: Just make sure to add `Telegram::listen(true);` in your webhook route. `true` in this case means to process inbound updates via webhook.
> 
> > Q2: In the event class, how do I send information and pictures? When I use $this->bot, it prompts me that I cannot find the sendPhoto method. Telegram::sendPhoto returns a null value, and here I need to know his message_id
> 
> If you're not using multiple bots, then you can simply do `Telegram::sendPhoto($params)`, just make sure to import the Telegram facade. If you're using multiple bots, then a better way to do it is using the Bot instance in the event.
> 
> Ex:
> 
> ```
> $bot = $event->bot;
> $bot->sendPhoto($params);
> ```
> 
> Don't worry about not found IDE warnings as these methods are being called through the magic methods dynamically linked from the Api class.
> 
> > Q3: 4. When can the documentation of X be updated...Using it without documentation will greatly affect development efficiency...
> 
> 4.x docs are being worked on, there are no ETAs as we've been refactoring a lot of things internally like the new `command()` methods and the new response objects and so on.



> Reply Q2
```
$bot = $event->bot;
$response = $bot->sendPhoto($params);
```

Brother, I used it this way today. '$response' is an empty array and cannot obtain the body of the 'sendPhoto' response. However, I noticed that you updated the 'decodeBody' function today. Is this previously empty array response due to a bug? After the update, I haven't tested according to this method yet

**Reply by irazasyed** [reply]
@HelplessMan Share your full code.

**Reply by HelplessMan** [reply]
>NewChatMember.php
```
<?php

namespace App\Telegram\Commands;

use Telegram\Bot\Commands\Command;
use Illuminate\Support\Facades\Cache;
use Telegram\Bot\Events\UpdateEvent;
use App\Common\Utils;
use Throwable;

class NewChatMember extends Command
{
    protected string $description = 'Command Description';

    private string $memberId;

    private string $chatId;

    /**
     * Execute the bot command.
     */
    public function handle(UpdateEvent $updateEvent)
    {
        logger('New chat member event: ' . json_encode($updateEvent));
        $this->chatId   = $updateEvent->update->message->chat->id;
        $this->memberId = $updateEvent->update->message->new_chat_member->id ?? '';
        $captchaContent = $this->getCaptchaContent();

        $key  = 'captcha_' . md5($captchaContent['key']);
        $text = implode('', Cache::get($key));

        $captchaUrl = $this->getCaptchaUrl($captchaContent);

        logger('message:' . json_encode($updateEvent, JSON_THROW_ON_ERROR) . ' text:' . $text . ' newMemberId:' . $this->memberId);

        logger('captchaUrl:' . $captchaUrl);
        //todo custom content
        $caption = 'Please click on the correct verification code to verify that you are a real person.';
        $response = $updateEvent->bot->sendPhoto([
            'chat_id' => $this->chatId,
            'photo' => $captchaUrl,
            'caption' => $caption,
            'reply_markup' => json_encode([
                'inline_keyboard' => Utils::generateCaptchaButton($text)
            ])
        ]);

        logger('sendPhoto ResponseObject:' . $response);
    }

    /**
     * Triggered on failure.
     *
     * @param array $arguments
     * @param Throwable $exception
     */
    public function failed(array $arguments, Throwable $exception): void
    {
        logger('Try new chat member join group exception:' . $exception->getMessage() .' '.  $exception->getFile() . ' params:' . json_encode($arguments, JSON_THROW_ON_ERROR));
    }

    public function getCaptchaContent(): array
    {
        //Send Http Request get generate captcha json params 
        $content = file_get_contents(Utils::captchaApi());

        if (empty($content)) {
            return [];
        }

        logger('Current captcha response content (chatId: ' . $this->chatId . ', memberId: ' . $this->memberId . '): ' . $content);

        return json_decode($content, true, 512, JSON_THROW_ON_ERROR);
    }

    public function getCaptchaUrl($captchaContent = []): string
    {
        if (isset($captchaContent['img'])) {

            $dir      = '/captcha';
            $path     = public_path() . $dir;
            $fileName = $this->memberId . '_' . now()->format('YmdHis');

            Utils::base64imageSave($captchaContent['img'] ?? '', $path, $fileName);

            $captchaUrl = config('app.url') . $dir . '/' . $fileName . '.png';

            logger('Captcha url (chatId: ' . $this->chatId . ', memberId: ' . $this->memberId . '):' . $captchaUrl);

            return $captchaUrl;
        }
        return '';
    }
}

```

> laravel.log
```
[2023-05-11 05:12:22] local.DEBUG: Captcha url (chatId: -1001784988088, memberId: 6273644092):https://api.telegramchat.in/captcha/5232644092_20230511051222.png  
[2023-05-11 05:12:22] local.DEBUG: message:{"bot":{},"update":{"update_id":363700312,"message":{"message_id":145,"from":{"id":5232644092,"is_bot":false,"first_name":"Kim","username":"Kim886686","language_code":"ms"},"chat":{"id":-1001843988088,"title":"test_join_verify_new","username":"test join","type":"supergroup"},"date":1683781942,"new_chat_participant":{"id":5232644092,"is_bot":false,"first_name":"Kim","username":"Kim886","language_code":"ms"},"new_chat_member":{"id":5232644092,"is_bot":false,"first_name":"Kim","username":"Kim886","language_code":"ms"},"new_chat_members":[{"id":5232644092,"is_bot":false,"first_name":"Kim","username":"Kim886","language_code":"ms"}]}}} text:jqgcze newMemberId:5232644092  
[2023-05-11 05:12:22] local.DEBUG: captchaUrl:https://api.telegramchat.in/captcha/5232644092_20230511051222.png  
[2023-05-11 05:12:22] local.DEBUG: sendPhoto ResponseObject:[]  
```

> group

<img width="580" alt="image" src="https://github.com/telegram-bot-sdk/telegram-bot-sdk/assets/9101275/2cc4d36e-2b57-4140-8bdb-d17ea5740f94">



@irazasyed  What I am doing now is the same function as the `PHP Chat` robot, verifying whether the newly added user is a bot, how to delete the current verification message after the verification is successful, because after calling `$updateEvent -> bot->sendPhoto` to send the information, Unable to get response object

**Reply by irazasyed** [reply]
Why is your listener a command? A listener is simple class with handle method, there is no need to extend, add description or all the things you're doing there.

**Reply by irazasyed** [reply]
I just tested this listener and it works perfectly fine:

```php
<?php

namespace App\Listeners;

use Telegram\Bot\Events\UpdateEvent;

class NewChatMember
{
    public function handle(UpdateEvent $event)
    {
        $update = $event->update;
        $bot = $event->bot;

        $photo = $bot->sendPhoto([
            'chat_id' => $update->message->chat->id,
            'photo' => 'https://telegram.org/img/t_logo.png',
        ]);
        
        $messageId = $photo['message_id'];
    }
}
```

**Reply by HelplessMan** [reply]
Sorry..

I modified it according to your method, but the feedback is still the same as my wrong use above, after calling `sendPhoto`, it still responds to an empty object....



> NewChatMember.php 
```
<?php

namespace App\Listeners;


use Illuminate\Support\Facades\Cache;
use Telegram\Bot\Events\UpdateEvent;
use App\Common\Utils;
use Throwable;

class NewChatMember
{
    private string $memberId;

    private string $chatId;

 
    public function handle(UpdateEvent $updateEvent)
    {
        logger('New chat member event: ' . json_encode($updateEvent));
        $this->chatId   = $updateEvent->update->message->chat->id;
        $this->memberId = $updateEvent->update->message->new_chat_member->id ?? '';
        $captchaContent = $this->getCaptchaContent();

        $key  = 'captcha_' . md5($captchaContent['key']);
        $text = implode('', Cache::get($key));

        $captchaUrl = $this->getCaptchaUrl($captchaContent);

        logger('message:' . json_encode($updateEvent, JSON_THROW_ON_ERROR) . ' text:' . $text . ' newMemberId:' . $this->memberId);

        logger('captchaUrl:' . $captchaUrl);
        //todo custom content
        $caption = 'Please click on the correct verification code to verify that you are a real person.';
        $response = $updateEvent->bot->sendPhoto([
            'chat_id' => $this->chatId,
            'photo' => $captchaUrl,
            'caption' => $caption,
            'reply_markup' => json_encode([
                'inline_keyboard' => Utils::generateCaptchaButton($text)
            ])
        ]);

        logger('sendPhoto ResponseObject:' . $response);
    }


    public function getCaptchaContent(): array
    {
        //Send Http Request get generate captcha json params 
        $content = file_get_contents(Utils::captchaApi());

        if (empty($content)) {
            return [];
        }

        logger('Current captcha response content (chatId: ' . $this->chatId . ', memberId: ' . $this->memberId . '): ' . $content);

        return json_decode($content, true, 512, JSON_THROW_ON_ERROR);
    }

    public function getCaptchaUrl($captchaContent = []): string
    {
        if (isset($captchaContent['img'])) {

            $dir      = '/captcha';
            $path     = public_path() . $dir;
            $fileName = $this->memberId . '_' . now()->format('YmdHis');

            Utils::base64imageSave($captchaContent['img'] ?? '', $path, $fileName);

            $captchaUrl = config('app.url') . $dir . '/' . $fileName . '.png';

            logger('Captcha url (chatId: ' . $this->chatId . ', memberId: ' . $this->memberId . '):' . $captchaUrl);

            return $captchaUrl;
        }
        return '';
    }
}
```

>telegram.php

```
'bots' => [
        'default' => [
            'token' => env('TELEGRAM_BOT_TOKEN', 'YOUR-BOT-TOKEN'),

            'commands' => [
                 'start' => App\Telegram\Commands\Start::class,
                 'info' => App\Telegram\Commands\Info::class,
                 'test' => App\Telegram\Commands\TestCommand::class,
            ],

            'listen' => [
                'update' => [],
                'webhook.failed' => [],

                // Example of various events fired.
                'message' => [
                    \App\Listeners\Message::class
                ],
                'poll' => [

                ],
                'message.left_chat_member' => [
                    \App\Listeners\LeftChatMember::class,
                ],
                'message.new_chat_members' => [
                    \App\Listeners\NewChatMember::class
                ],
                'inline_query.location' => [],
                'callback_query' => [
                    \App\Listeners\CallbackQuery::class,
                ]
            ],
        ],
]
```

@irazasyed  I execute the demo listener, the response is still empty

```
<?php

namespace App\Listeners;

use Telegram\Bot\Events\UpdateEvent;
use Telegram\Bot\Exceptions\TelegramSDKException;
use Telegram\Bot\Objects\PhotoSize;

/**
 * Class ProcessInboundPhoto
 */
class ProcessInboundPhoto
{
    /**
     * Handle the event.
     *
     *
     * @throws TelegramSDKException
     */
    public function handle(UpdateEvent $event)
    {
        $update = $event->update;
        $bot = $event->bot;

        // Reply the user.
        $text = 'Thanks for uploading the pic!';
        $response = $bot->sendMessage([
            'chat_id' => $update->message->chat->id,
            'text' => $text,
        ]);

        $messageId = $response['message_id'];

        logger('test photo' . $messageId);
    }
}
```

**Reply by HelplessMan** [reply]
@irazasyed  Brother, remember what I mentioned the day before yesterday, in the listener, after sending a message, you can’t get the response object? I have now found the reason for this problem. It is because I set `TELEGRAM_ASYNC_REQUESTS=true` in .so the response object will not be received. If I set `TELEGRAM_ASYNC_REQUESTS=false`, I can get the response object.

**Reply by irazasyed** [reply, marked as answer]
> @irazasyed  Brother, remember what I mentioned the day before yesterday, in the listener, after sending a message, you can’t get the response object? I have now found the reason for this problem. It is because I set `TELEGRAM_ASYNC_REQUESTS=true` in .so the response object will not be received. If I set `TELEGRAM_ASYNC_REQUESTS=false`, I can get the response object.

Ahan. Good find 👍🏻

Yes. In async mode you don't get the response. Hopefully everything is working fine now for you.

**Comment by irazasyed** [comment]
To setup webhook, make sure you've setup this env var:

`TELEGRAM_WEBHOOK_DOMAIN=mydomain.com` (replace with your own domain).

Then run `php artisan telegram:webhook:setup`

Keep in mind, you need to install the `dev-main` version of the Laravel package as it's not yet released.

**Reply by irazasyed** [reply]
You also need to keep in mind your domain already has SSL enabled, i.e. `https` is supported.

When you setup the webhook, you don't need to manually listen or do anything in Laravel. The Laravel package ships with the webhook controller and it's already setup to handle multiple bots for you. 

You can get started developing your bots by listening to inbound events and setting up commands.

**Reply by HelplessMan** [reply]
Brother, I am currently using the 'dev main' version and I really like this method. This is really great, but due to the lack of documentation at the moment, I need to read the SDK source code to solve some problems, which is a bit painful. I will summarize the problems I have encountered here and hope to help others in the future

**Reply by irazasyed** [reply]
Unfortunately 4.x is a WIP version and is still in the works which is why it's taking longer than we anticipated to work on the docs.

You are however welcome to contribute to the project if you've learnt from the source code to our docs.

**Reply by HelplessMan** [reply]
Of course, thank you for contributing such a useful telegram bot sdk
